File: Data Structures & Algorithms/search-2d-matrix/submission-1.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        '''
        m, n = len(matrix), len(matrix[0])
        l, r = 0, m * n - 1
        while l <= r:
            mid = (l + r) // 2
            row = mid // n # Tìm vị trí của hàng vì mỗi hàng chứa n phần tử, ví dụ mỗi hàng có 4 phần tử thì biết mid / 4 sẽ ra vị trí hàng
            col = mid % n #Phép chia dư tìm vị trí cột trong hàng
            if matrix[row][col] == target:
                return True
            elif matrix[row][col] < target:
                l = mid + 1
            else:
                r = mid - 1
        return False
        '''
        m, n = len(matrix), len(matrix[0])
        l, r = 0, m * n - 1
        def binary_search(l: int, r: int) -> bool:
            if l > r:
                return False
            mid = (l + r) // 2
            row = mid // n
            col = mid % n
            if matrix[row][col] == target:
                return True
            elif matrix[row][col] < target:
                return binary_search(mid + 1, r)
            else:
                return binary_search(l, mid - 1)
        return binary_search(l, r)
            

File: Data Structures & Algorithms/max-area-of-island/submission-1.cpp
class Solution {
public:
    int rows;
    int cols;

    int dfs(vector<vector<int>>& grid, int row, int col) {
        //Ngoài vùng grid
        if (row < 0 || row >= rows || col < 0 || col >= cols) return 0;
        // Ô là nước
        if (grid[row][col] == 0) return 0;
        //Đánh dấu ô hiện tại đã được thăm
        grid[row][col] = 0;
        //Đếm số ô của 1 vùng island theo 4 hướng
        return 1 
            + dfs(grid, row - 1, col) 
            + dfs(grid, row + 1, col) 
            + dfs(grid, row, col - 1) 
            + dfs(grid, row, col + 1);
    }

    int maxAreaOfIsland(vector<vector<int>>& grid) {
        this->rows = grid.size();
        this->cols = grid[0].size();
        int maxArea = 0;
        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                if (grid[row][col] == 1) {
                    int area = dfs(grid, row, col);
                    maxArea = max(maxArea, area);
                }
            }
        }
        return maxArea;
    }
};

/*
TimeCP: O(n * m); n là số hàng và m là số cột vì phải duyệt qua từng ô trong grid
SpaceCP: O(n * m) in worst case bởi vì độ sâu đệ quy lớn nhất là 1 island lớn full 1
*/